src/extender.test.js

```diff
@@ -0,0 +1,204 @@
+    shimmer.unwrap(DummyCounterService, 'incrementToDummyCounterMultipleParam');
+    shimmer.unwrap(DummyCounterService, 'getDummyCounter');
+    DummyCounterService.reset();
+  });
+
+  test('hook -> simple flow, empty before and after', () => {
+    extender.hook(DummyCounterService, 'incrementToDummyCounter');
+    DummyCounterService.incrementToDummyCounter();
+    expect(DummyCounterService.getDummyCounter()).toEqual(1);
+  });
+
+  test('hook -> before Hook', () => {
+    let beforeCounter = 0;
+    extender.hook(DummyCounterService, 'incrementToDummyCounter', {
+      beforeHook: () => {
+        expect(DummyCounterService.getDummyCounter()).toEqual(0);
+        beforeCounter++;
+      },
+    });
+    DummyCounterService.incrementToDummyCounter();
+    expect(DummyCounterService.getDummyCounter()).toEqual(1);
+    expect(beforeCounter).toEqual(1);
+  });
+
+  test('hook -> before Hook - multiple param', () => {
+    let beforeCounter = 0;
+    extender.hook(DummyCounterService, 'incrementToDummyCounterMultipleParam', {
+      beforeHook: args => {
+        expect(DummyCounterService.getDummyCounter()).toEqual(0);
+        beforeCounter += args[0];
+        beforeCounter += args[1];
+      },
+    });
+    DummyCounterService.incrementToDummyCounterMultipleParam(5, 10);
+    expect(DummyCounterService.getDummyCounter()).toEqual(15);
+    expect(beforeCounter).toEqual(15);
+  });
+
+  test('hook -> after Hook', () => {
+    let afterCounter = 0;
+    extender.hook(DummyCounterService, 'incrementToDummyCounter', {
+      afterHook: () => {
+        expect(DummyCounterService.getDummyCounter()).toEqual(1);
+        afterCounter++;
+      },
+    });
+    DummyCounterService.incrementToDummyCounter();
+    expect(DummyCounterService.getDummyCounter()).toEqual(1);
+    expect(afterCounter).toEqual(1);
+  });
+
+  test('hook -> before & after Hook', () => {
+    let afterCounter = 0;
+    let beforeCounter = 0;
+    extender.hook(DummyCounterService, 'incrementToDummyCounter', {
+      beforeHook: () => {
+        expect(DummyCounterService.getDummyCounter()).toEqual(0);
+        beforeCounter++;
+      },
+      afterHook: () => {
+        expect(DummyCounterService.getDummyCounter()).toEqual(1);
+        afterCounter++;
+      },
+    });
+    DummyCounterService.incrementToDummyCounter();
+    expect(DummyCounterService.getDummyCounter()).toEqual(1);
+    expect(afterCounter).toEqual(1);
+    expect(beforeCounter).toEqual(1);
+  });
+
+  test('hook -> extenderContext', () => {
+    let afterCounter = 0;
+    let beforeCounter = 0;
+    extender.hook(DummyCounterService, 'incrementToDummyCounter', {
+      beforeHook: (args, extenderContext) => {
+        expect(extenderContext).toEqual({});
+        extenderContext.a = '1';
+        beforeCounter++;
+      },
+      afterHook: (args, functionResult, extenderContext) => {
+        expect(extenderContext).toEqual({ a: '1' });
+        afterCounter++;
```

Comment:
here too - `done` is the better way to go 🤷

Reply:
And how you can make sure it called after the user function?

Reply:
`expect(DummyCounterService.getDummyCounter()).toEqual(1);`
?